src/docmanager/index.ts

```diff
@@ -0,0 +1,514 @@
+
+  /**
+   * Serialize the model.  It should return a JSON object or a string.
+   */
+  serialize(): any;
+
+  /**
+   * Deserialize the model from a string or a JSON object.
+   *
+   * #### Notes
+   * Should emit a [contentChanged] signal.
+   */
+  deserialize(value: any): void;
+
+  /**
+   * The default kernel name for the document.
+   *
+   * #### Notes
+   * This is a read-only property.
+   */
+  defaultKernelName: string;
+
+  /**
+   * The default kernel language for the document.
+   *
+   * #### Notes
+   * This is a read-only property.
+   */
+  defaultKernelLanguage: string;
+}
+
+
+/**
+ * A session info object for running sessions.
+ */
+export interface ISessionInfo {
+  /**
+   * The list of file paths associated with the running sessions.
+   */
+  paths: string[];
+
+  /**
+   * The list of kernel specs associated with the running sessions.
+   */
+  specs: IKernelSpecId[];
+}
+
+
+/**
+ * The document context object.
+ */
+export interface IDocumentContext {
+  /**
+   * A signal emitted when the kernel changes.
+   */
+  kernelChanged: ISignal<IDocumentContext, IKernel>;
+
+  /**
+   * A signal emitted when the path changes.
+   */
+  pathChanged: ISignal<IDocumentContext, string>;
+
+  /**
+   * A signal emitted when the model is saved or reverted.
+   */
+  dirtyCleared: ISignal<IDocumentContext, void>;
+
+  /**
+   * The current kernel associated with the document.
+   *
+   * #### Notes
+   * This is a read-only property.
+   */
+  kernel: IKernel;
+
+  /**
+   * The current path associated with the document.
+   *
+   * #### Notes
+   * This is a read-only property.
+   */
+  path: string;
+
+  /**
+   * Change the current kernel associated with the document.
+   */
+  changeKernel(options: IKernelId): Promise<IKernel>;
+
+  /**
+   * Save the document contents to disk.
+   */
+  save(): Promise<void>;
+
+  /**
+   * Revert the document contents to disk contents.
+   */
+  revert(): Promise<void>;
+
+  /**
+   * Get the list of running sessions.
+   */
+  listRunningSessions(): Promise<ISessionInfo>;
+
+  /**
+   * Add a sibling widget to the document manager.
+   *
+   * @param widget - The widget to add to the document manager.
+   *
+   * @returns A disposable used to remove the sibling if desired.
+   */
+  addSibling(widget: Widget): IDisposable;
+}
+
+
+/**
+ * The interface for a widget factory.
+ */
+export
+interface IWidgetFactory<T extends Widget> {
+  /**
+   * Create a new widget given a document model and a context.
+   */
+  createNew(model: IDocumentModel, context: IDocumentContext): T;
+
+  /**
+   * Get the default kernel name given a path and a list of specs.
+   */
+  getDefaultKernel(path: string, specs: IKernelSpecId[]): string;
+
+  /**
+   * Get the preferred kernel names given a path and a list of specs.
+   */
+  getPreferredKernels(path: string, specs: IKernelSpecId[]): string[];
+
+  /**
+   * Get the preferred widget title given a path.
+   */
+  getWidgetTitle(path: string): string;
```

Comment:
@blink1073 - however, we might want to put widget-specific state in the title. For example, a star indicating a dirty document, or a badge that indicates unread messages, etc. So maybe this method should also take the widget instance.